### PR TITLE
Basic Metal BRDF

### PR DESCRIPTION
**Pull request by est77**
Quick render: gold teapot

metal

### CODE REVIEW

**Comment by est77** [comment]
Quick render: gold teapot

metal

**Comment by Narann** [comment]
Very nice!

**Comment by dictoon** [comment]
Very nice indeed!

File: src/appleseed/foundation/math/fresnel.h [hunk 1]
```diff
@@ -130,6 +130,57 @@ void normal_reflectance_dielectric(
 
 
 //
+// Fresnel equations for conductor materials.
+//
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Remove one blank line.

File: src/appleseed/foundation/math/fresnel.h [hunk 2]
```diff
@@ -130,6 +130,57 @@ void normal_reflectance_dielectric(
...
+//
+// References:
+//
+//   [1] Physically Based Rendering, first edition.
+//   [2] Artist Friendly Metallic Fresnel.
+//       Ole Gulbrandsen
+//       http://jcgt.org/published/0003/04/03/paper.pdf
+//
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Remove one blank line.

File: src/appleseed/foundation/math/fresnel.h [hunk 3]
```diff
@@ -130,6 +130,57 @@ void normal_reflectance_dielectric(
...
+//
+
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Remove one blank line.

File: src/appleseed/foundation/math/fresnel.h [hunk 4]
```diff
@@ -130,6 +130,57 @@ void normal_reflectance_dielectric(
...
+
+// Compute the Fresnel reflectance for a conductor for unpolarized light.
+// This approximation assumes an air - conductor interface.
+template <typename SpectrumType, typename T>
+void fresnel_reflectance_conductor(
+    SpectrumType&           reflectance,
+    const SpectrumType&     n,
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Add comments (on the same lines) to explain what `n` and `k` are?

File: src/appleseed/foundation/math/fresnel.h [hunk 5]
```diff
@@ -383,6 +434,138 @@ void normal_reflectance_dielectric(
...
+    //               Rp + Rs
+    // reflectance = -------
+    //                  2
+
+    typedef typename impl::GetValueType<SpectrumType>::ValueType ValueType;
+
+    const T cos_theta = std::abs(cos_theta_i);
+    const T cos_theta2 = square(cos_theta);
+
+    const SpectrumType n2 = n * n;
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Not sure how this compares in terms of generated code to:

    SpectrumType n2(n);
    n2 *= n2;

File: src/appleseed/foundation/math/fresnel.h [hunk 6]
```diff
@@ -383,6 +434,138 @@ void normal_reflectance_dielectric(
...
+    const SpectrumType&     normal_reflectance,
+    const SpectrumType&     edge_tint,
+    SpectrumType&           n,
+    SpectrumType&           k)
+{
+    typedef typename impl::GetValueType<SpectrumType>::ValueType ValueType;
+
+    const SpectrumType one(1.0);
+    const SpectrumType r = clamp(normal_reflectance, ValueType(0.0), ValueType(0.99));
+    const SpectrumType sqrt_r = sqrt(r);
```
Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Missing `std::` qualifier.

File: src/appleseed/foundation/meta/tests/test_fresnel.cpp [hunk 7]
```diff
@@ -321,4 +321,211 @@ TEST_SUITE(Foundation_Math_Fresnel)
         fresnel_transmittance_dielectric(tr, Eta, Eps); EXPECT_FEQ(0.0, tr);
         fresnel_transmittance_dielectric(tr, Eta, 1.0); EXPECT_EQ(1.0, tr);
     }
+
+    // Source: http://refractiveindex.info
+    // Wavelenghts: 645, 526 and 444 nm.
```
Comment (dictoon) [inline src/appleseed/foundation/meta/tests/test_fresnel.cpp]: "Wavelengths"

File: src/appleseed/renderer/modeling/bsdf/metalbrdf.cpp [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/metalbrdf.cpp]: Remove redundant blank line.

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 9]
```diff
@@ -976,6 +982,24 @@ inline renderer::DynamicSpectrum<T, N> clamp_high(const renderer::DynamicSpectru
...
+    const renderer::DynamicSpectrum<T, N>& t)
+{
+    assert(a.size() == b.size());
+    assert(a.size() == t.size());
+
+    renderer::DynamicSpectrum<T, N> result;
+    result.resize(a.size());
+
+    for (size_t i = 0, e = a.size(); i < e; ++i)
+        result[i] = (a[i] * (T(1) - t[i])) + (b[i] * t[i]);
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h]: Use `foundation::lerp()`?

File: src/appleseed/foundation/math/fresnel.h (hunk 6 above)
Comment (est77) [inline src/appleseed/foundation/math/fresnel.h]: I need to use argument dependent lookup here to pick sqrt depending on SpectrumType.
std::sqrt when SpectrumType is double (in unit tests) and renderer::sqrt when 
SpectrumType is a DynamicSpectrum (when rendering).

Comment (dictoon) [inline src/appleseed/foundation/math/fresnel.h]: Good point.

**Pull request merged**